File: crates/buildbridge-machines/src/transporter.py
"""Receive one retained IPA and a one-use App Store Connect credential over SSH."""

import hashlib
import os
import re
import select
import signal
import struct
import subprocess
import sys
import tempfile
import threading
import time


UPLOAD_TIMEOUT_SECONDS = 7200
PREFIX = "__BUILDBRIDGE_TRANSPORTER__:"


class UploadError(Exception):
    pass


def report(message):
    print(PREFIX + message, flush=True)


def read_exact(stream, length):
    chunks = bytearray()
    while len(chunks) < length:
        chunk = stream.read(length - len(chunks))
        if not chunk:
            raise UploadError("The IPA transfer was interrupted; no upload was started.")
        chunks.extend(chunk)
    return bytes(chunks)


def read_frame(stream, maximum):
    length = struct.unpack(">I", read_exact(stream, 4))[0]
    if length > maximum:
        raise UploadError("An upload credential exceeds its size limit.")
    return read_exact(stream, length).decode("utf-8")


def find_transporter():
    applications = [os.path.expanduser("~/Applications"), "/Applications"]
    candidates = []
    for directory in applications:
        candidates.append(directory + "/Transporter.app/Contents/itms/bin/iTMSTransporter")
        candidates.append(directory + "/Xcode.app/Contents/SharedFrameworks/ContentDeliveryServices.framework/Versions/A/itms/bin/iTMSTransporter")
        candidates.append(directory + "/Xcode.app/Contents/Applications/Application Loader.app/Contents/itms/bin/iTMSTransporter")
    candidates.append("/usr/local/itms/bin/iTMSTransporter")
    try:
        developer = subprocess.check_output(
            ["/usr/bin/xcode-select", "--print-path"],
            stderr=subprocess.DEVNULL,
            timeout=5,
        ).decode("utf-8").strip()
        candidates.append(os.path.dirname(developer) + "/SharedFrameworks/ContentDeliveryServices.framework/Versions/A/itms/bin/iTMSTransporter")
    except (OSError, subprocess.SubprocessError, UnicodeError):
        pass
    for candidate in candidates:
        if os.path.isfile(candidate) and os.access(candidate, os.X_OK):
            return candidate
    raise UploadError("Transporter is not installed in the macOS machine. Install Apple's Transporter app from the Mac App Store in that machine, then retry.")


def stop_process(process):
    # Transporter's launcher starts Java children. Stop the whole session before
    # removing the key and IPA, including when the SSH connection disappears.
    if process is None:
        return
    try:
        os.killpg(process.pid, signal.SIGTERM)
    except ProcessLookupError:
        return
    try:
        process.wait(timeout=2)
    except subprocess.TimeoutExpired:
        pass
    try:
        os.killpg(process.pid, signal.SIGKILL)
    except ProcessLookupError:
        pass
    process.wait(timeout=3)


def summarize_output(stream, summary):
    # Never forward Transporter's raw output, which may contain authentication
    # material. Keep only a small window for classification and numeric ITMS codes.
    window = b""
    while True:
        chunk = stream.read(4096)
        if not chunk:
            return
        window = (window + chunk)[-8192:]
        lower = window.lower()
        if any(value in lower for value in (b"unauthorized", b"authentication", b"invalid jwt", b"not authorized")):
            summary["authentication"] = True
        if any(value in lower for value in (b"already been used", b"duplicate binary", b"redundant binary", b"itms-90189", b"itms-90186")):
            summary["duplicate"] = True
        for code in re.findall(rb"\bITMS-[0-9]{4,6}\b", window):
            if len(summary["codes"]) < 10:
                summary["codes"].add(code)


def disconnected(stream):
    ready, _, _ = select.select([stream], [], [], 0)
    if not ready:
        return False
    # After the payload the host retains stdin without sending anything. EOF
    # means SSH was cancelled or disconnected; extra bytes are also invalid.
    stream.read(1)
    return True


def upload(stream):
    transporter = find_transporter()
    key_id = read_frame(stream, 64)
    issuer_id = read_frame(stream, 64)
    private_key = read_frame(stream, 16384)
    expected_sha256 = read_frame(stream, 64)
    size = struct.unpack(">Q", read_exact(stream, 8))[0]
    if not re.fullmatch(r"[A-Z0-9]{10}", key_id):
        raise UploadError("Enter the 10-character App Store Connect API Key ID.")
    if not re.fullmatch(r"[0-9a-fA-F]{8}(?:-[0-9a-fA-F]{4}){3}-[0-9a-fA-F]{12}", issuer_id):
        raise UploadError("Enter the App Store Connect team API key's Issuer ID.")
    if not re.fullmatch(r"[0-9a-fA-F]{64}", expected_sha256) or not 0 < size <= 16 * 1024**3:
        raise UploadError("The retained IPA has invalid size or checksum metadata.")
    if not private_key.startswith("-----BEGIN PRIVATE KEY-----") or not private_key.rstrip().endswith("-----END PRIVATE KEY-----"):
        raise UploadError("Choose the App Store Connect API private key (.p8).")

    os.umask(0o077)
    with tempfile.TemporaryDirectory(prefix="buildbridge-transporter-", dir="/tmp") as directory:
        ipa = os.path.join(directory, "App-AppStore.ipa")
        remaining = size
        digest = hashlib.sha256()
        with open(ipa, "xb") as target:
            while remaining:
                chunk = read_exact(stream, min(remaining, 256 * 1024))
                target.write(chunk)
                digest.update(chunk)
                remaining -= len(chunk)
        if digest.hexdigest() != expected_sha256.lower():
            raise UploadError("The IPA changed during transfer. Its checksum does not match the retained release; no upload was started.")
        if disconnected(stream):
            raise UploadError("Upload stopped before delivery to Apple.")
        keys = os.path.join(directory, "private_keys")
        os.mkdir(keys, 0o700)
        with open(os.path.join(keys, "AuthKey_" + key_id + ".p8"), "x") as key:
            key.write(private_key)

        process = None
        try:
            report("The retained IPA is verified. Transporter is uploading to App Store Connect.")
            process = subprocess.Popen(
                [transporter, "-m", "upload", "-assetFile", ipa, "-apiKey", key_id, "-apiIssuer", issuer_id, "-v", "critical"],
                cwd=directory,
                stdin=subprocess.DEVNULL,
                stdout=subprocess.PIPE,
                stderr=subprocess.STDOUT,
                start_new_session=True,
            )
            summary = {"authentication": False, "duplicate": False, "codes": set()}
            reader = threading.Thread(target=summarize_output, args=(process.stdout, summary), daemon=True)
            reader.start()
            while process.poll() is None:
                if disconnected(stream):
                    raise UploadError("Upload stopped. Check App Store Connect before retrying; Apple may have received the build.")
                time.sleep(0.2)
            reader.join(timeout=2)
            if disconnected(stream):
                raise UploadError("Upload stopped. Check App Store Connect before retrying; Apple may have received the build.")
            if process.returncode:
                if summary["authentication"]:
                    raise UploadError("Transporter could not authenticate. Check the team API key, Issuer ID, and the key's access to this app in App Store Connect.")
                if summary["duplicate"]:
                    raise UploadError("Apple has already received this version or build number. Check the existing build in App Store Connect, or archive a new build with a higher build number.")
                codes = ", ".join(code.decode("ascii") for code in sorted(summary["codes"])[:10])
                detail = (" Apple reported " + codes + ".") if codes else ""
                raise UploadError("Transporter rejected the upload (exit " + str(process.returncode) + ")." + detail + " Check this app's bundle ID, version/build number and account access in App Store Connect before retrying.")
        finally:
            stop_process(process)
    report("Upload delivered. Apple must process the build before it appears in TestFlight.")


def stopped(signum, frame):
    if signum == signal.SIGALRM:
        raise UploadError("The upload timed out. Check App Store Connect before retrying; Apple may have received the build.")
    raise UploadError("Upload stopped. Check App Store Connect before retrying; Apple may have received the build.")


def main():
    for signum in (signal.SIGHUP, signal.SIGTERM, signal.SIGINT, signal.SIGALRM):
        signal.signal(signum, stopped)
    signal.alarm(UPLOAD_TIMEOUT_SECONDS)
    try:
        # Unbuffered input is essential: select must observe EOF only after the
        # exact payload, while the host holds the connection open for completion.
        with os.fdopen(os.dup(sys.stdin.fileno()), "rb", buffering=0) as stream:
            upload(stream)
    except UploadError as error:
        report("ERROR: " + str(error))
        return 1
    except (OSError, ValueError, UnicodeError, subprocess.SubprocessError):
        report("ERROR: The macOS upload could not finish. Check the machine's connection and available storage, then retry.")
        return 1
    finally:
        signal.alarm(0)
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: crates/buildbridge-machines/src/transporter.rs
//! Deliver the exact retained IPA through Transporter inside the pinned macOS guest.

use super::*;
use std::sync::mpsc;

const TRANSPORTER_HELPER: &str = include_str!("transporter.py");
const TRANSPORTER_PREFIX: &str = "__BUILDBRIDGE_TRANSPORTER__:";
const TRANSPORTER_COMPLETED: &str =
    "Upload delivered. Apple must process the build before it appears in TestFlight.";
const TRANSPORTER_TIMEOUT: Duration = Duration::from_secs(2 * 60 * 60 + 15);

/// Uses an installed Transporter with a team App Store Connect API key. The key is
/// carried through framed stdin and removed from the guest on completion, failure,
/// timeout, or disconnection. The retained IPA must match `expected_sha256` after
/// transfer before any delivery to Apple can begin.
#[allow(clippy::too_many_arguments)]
pub fn upload_apple_ipa<F>(
    ssh_port: u16,
    username: &str,
    identity_path: &Path,
    known_hosts_path: &Path,
    ipa_path: &Path,
    expected_sha256: &str,
    key_id: &str,
    issuer_id: &str,
    private_key: &str,
    mut on_progress: F,
) -> Result<(), ProviderError>
where
    F: FnMut(String),
{
    validate_upload_input(username, expected_sha256, key_id, issuer_id, private_key)?;
    let ipa = File::open(ipa_path).map_err(|error| {
        upload_error(format!(
            "could not open the retained App Store IPA: {error}"
        ))
    })?;
    let metadata = ipa.metadata().map_err(|error| {
        upload_error(format!(
            "could not inspect the retained App Store IPA: {error}"
        ))
    })?;
    if !metadata.is_file() || metadata.len() == 0 || metadata.len() > 16 * 1024 * 1024 * 1024 {
        return Err(upload_error(
            "the retained App Store IPA is empty or exceeds 16 GiB",
        ));
    }
    let ssh = guest_ssh_command(ssh_port, username, identity_path, known_hosts_path);
    let mut command = Command::new(ssh.get_program());
    // All guest source is fixed; paths, identifiers and the private key are data on
    // stdin, never shell text. Keepalives bound a severed network connection.
    command
        .args([
            "-o",
            "ServerAliveInterval=15",
            "-o",
            "ServerAliveCountMax=3",
        ])
        .args(ssh.get_args())
        .arg(transporter_command());
    run_upload(
        &mut command,
        ipa,
        metadata.len(),
        [key_id, issuer_id, private_key, expected_sha256],
        TRANSPORTER_TIMEOUT,
        &mut on_progress,
    )
}

fn transporter_command() -> String {
    format!(
        "exec /usr/bin/python3 -u -c {}",
        shell_single_quote(TRANSPORTER_HELPER)
    )
}

fn upload_error(message: impl Into<String>) -> ProviderError {
    ProviderError::GuestBridge(message.into())
}

fn validate_upload_input(
    username: &str,
    expected_sha256: &str,
    key_id: &str,
    issuer_id: &str,
    private_key: &str,
) -> Result<(), ProviderError> {
    let uuid_parts = issuer_id.split('-').collect::<Vec<_>>();
    if !valid_guest_username(username) {
        return Err(upload_error("the macOS guest username is invalid"));
    }
    if expected_sha256.len() != 64
        || !expected_sha256
            .bytes()
            .all(|character| character.is_ascii_hexdigit())
    {
        return Err(upload_error(
            "the retained App Store IPA checksum is invalid",
        ));
    }
    if key_id.len() != 10
        || !key_id
            .bytes()
            .all(|character| character.is_ascii_uppercase() || character.is_ascii_digit())
    {
        return Err(upload_error(
            "enter the 10-character App Store Connect API Key ID",
        ));
    }
    if uuid_parts.len() != 5
        || uuid_parts
            .iter()
            .zip([8, 4, 4, 4, 12])
            .any(|(part, length)| {
                part.len() != length || !part.bytes().all(|character| character.is_ascii_hexdigit())
            })
    {
        return Err(upload_error(
            "enter the App Store Connect team API key's Issuer ID",
        ));
    }
    if private_key.len() > 16_384
        || !private_key.starts_with("-----BEGIN PRIVATE KEY-----")
        || !private_key
            .trim_end()
            .ends_with("-----END PRIVATE KEY-----")
        || private_key.contains('\0')
    {
        return Err(upload_error(
            "choose the App Store Connect API private key (.p8)",
        ));
    }
    Ok(())
}

enum UploadEvent {
    Output(String),
    Transfer(Result<(), ProviderError>),
}

fn write_upload_payload(
    writer: &mut impl Write,
    ipa: &mut impl Read,
    size: u64,
    frames: [&str; 4],
) -> Result<(), ProviderError> {
    for frame in frames {
        write_secret_frame(writer, frame)?;
    }
    writer.write_all(&size.to_be_bytes()).map_err(|error| {
        upload_error(format!("could not send the retained IPA length: {error}"))
    })?;
    let copied = std::io::copy(&mut ipa.take(size), writer).map_err(|error| {
        upload_error(format!(
            "the retained IPA transfer was interrupted: {error}"
        ))
    })?;
    if copied != size {
        return Err(upload_error(
            "the retained IPA changed while it was being transferred",
        ));
    }
    writer.flush().map_err(|error| {
        upload_error(format!(
            "could not finish the retained IPA transfer: {error}"
        ))
    })
}

fn read_upload_output(reader: impl Read, events: mpsc::Sender<UploadEvent>) {
    let mut reader = BufReader::new(reader);
    let mut line = Vec::new();
    loop {
        line.clear();
        // Bound each line before parsing. Only our helper's progress protocol is
        // exposed; Transporter stdout and stderr never reach the host logs.
        match reader.by_ref().take(4097).read_until(b'\n', &mut line) {
            Ok(0) | Err(_) => return,
            Ok(_) if line.len() > 4096 => return,
            Ok(_) => {
                let value = String::from_utf8_lossy(&line);
                if let Some(message) = value.trim().strip_prefix(TRANSPORTER_PREFIX) {
                    let _ = events.send(UploadEvent::Output(message.to_string()));
                }
            }
        }
    }
}

fn run_upload(
    command: &mut Command,
    mut ipa: File,
    size: u64,
    frames: [&str; 4],
    timeout: Duration,
    on_progress: &mut impl FnMut(String),
) -> Result<(), ProviderError> {
    let operation = current_scope();
    let mut child = command
        .stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .stderr(Stdio::null())
        .tracked_spawn()
        .map_err(|error| {
            upload_error(format!(
                "could not start the Transporter SSH bridge: {error}"
            ))
        })?;
    let mut stdin = child.stdin.take().expect("the upload stdin is piped");
    let stdout = child.stdout.take().expect("the upload stdout is piped");
    let started = Instant::now();
    on_progress("Transferring the retained App Store IPA to the macOS machine.".to_string());
    let (events, receiver) = mpsc::channel();
    let (release, hold_input) = mpsc::channel::<()>();
    let mut release = Some(release);
    thread::scope(|threads| {
        let transfer_events = events.clone();
        threads.spawn(move || {
            let result = write_upload_payload(&mut stdin, &mut ipa, size, frames);
            let failed = result.is_err();
            let _ = transfer_events.send(UploadEvent::Transfer(result));
            if !failed {
                // EOF is the guest's cancellation signal. Retain the pipe until
                // the guest has finished, or the host cancels/times out.
                let _ = hold_input.recv();
            }
        });
        let output_reader = threads.spawn(move || read_upload_output(stdout, events));
        let mut output_reader = Some(output_reader);
        let mut completed = false;
        let mut error = None;
        let result = loop {
            if operation.as_ref().is_some_and(|scope| scope.is_cancelled()) {
                break Err(upload_error(
                    "upload stopped; check App Store Connect before retrying because Apple may have received the build",
                ));
            }
            if started.elapsed() > timeout {
                break Err(upload_error(
                    "the upload timed out; check App Store Connect before retrying because Apple may have received the build",
                ));
            }
            for event in receiver.try_iter() {
                match event {
                    UploadEvent::Output(message) => {
                        if let Some(message) = message.strip_prefix("ERROR: ") {
                            error = Some(upload_error(message));
                        } else {
                            completed |= message == TRANSPORTER_COMPLETED;
                            on_progress(message);
                        }
                    }
                    UploadEvent::Transfer(Err(transfer_error)) => {
                        // A missing guest tool may close stdin before transfer
                        // finishes. Prefer the helper's actionable error.
                        if error.is_none() {
                            error = Some(transfer_error);
                        }
                    }
                    UploadEvent::Transfer(Ok(())) => {}
                }
            }
            match child.try_wait() {
                Ok(Some(status)) => {
                    // An ended SSH process can no longer confirm delivery. Stop
                    // retaining stdin before joining output so a surviving guest
                    // helper can observe EOF and remove its temporary key.
                    drop(release.take());
                    // Readers can still have the final output buffered after
                    // the process exits. Drain their final events before deciding.
                    let _ = output_reader
                        .take()
                        .expect("the output reader is joined once")
                        .join();
                    for event in receiver.try_iter() {
                        if let UploadEvent::Output(message) = event {
                            if let Some(message) = message.strip_prefix("ERROR: ") {
                                error = Some(upload_error(message));
                            } else {
                                completed |= message == TRANSPORTER_COMPLETED;
                                on_progress(message);
                            }
                        }
                    }
                    break if let Some(error) = error {
                        Err(error)
                    } else if status.success() && completed {
                        Ok(())
                    } else {
                        Err(upload_error(
                            "the Transporter connection ended without confirming delivery; verify macOS SSH and Xcode's Python tools, and check App Store Connect before retrying",
                        ))
                    };
                }
                Ok(None) => thread::sleep(Duration::from_millis(100)),
                Err(error) => {
                    break Err(upload_error(format!(
                        "could not monitor Transporter: {error}"
                    )));
                }
            }
        };
        drop(release);
        if child.try_wait().ok().flatten().is_none() {
            let _ = child.kill();
        }
        let _ = child.wait();
        if operation.as_ref().is_some_and(|scope| scope.is_cancelled()) {
            return Err(upload_error(
                "upload stopped; check App Store Connect before retrying because Apple may have received the build",
            ));
        }
        result
    })
}

#[cfg(all(test, unix))]
#[path = "transporter_tests.rs"]
mod tests;
